### ISSUE TITLE
[autodiff] Defining a local variable inside the inner loop raises runtime error in llvm codegen

### ISSUE THREAD

**Issue by erizmr** [taichi-dev/taichi#4038]
**Describe the bug**
In the autodiff system, with fixed `ad_stack_size` (i.e. skip the `determine_ad_stack_size`), defining a local variable inside the inner loop raises runtime error in llvm codegen.

**To Reproduce**

```py
import taichi as ti

ti.init(print_ir=True, ad_stack_size=32)
x = ti.field(dtype=float, shape=(), needs_grad=True)
arr = ti.field(dtype=float, shape=(2), needs_grad=True)
loss = ti.field(dtype=float, shape=(), needs_grad=True)


@ti.kernel
def ad_mixed_inner_loops():
    for a in range(2):
        # It seems that the local variable can only be defined in the 
        # scope of the outmost loop, otherwise, runtime error
        for i in range(3):
            s = 0.0
            for l in range(3):
                s += ti.sin(x[None])
            loss[None] += s
x[None] = 0.0
with ti.Tape(loss=loss):
    ad_mixed_inner_loops()

print(loss[None], x.grad[None])


```

**Log/Screenshots**

```
python3: /home/leon/repos/llvm-10.0.0.src/lib/IR/Instructions.cpp:1
356: void llvm::StoreInst::AssertOK():  Assertion `getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"' failed.

```

**Additional comments**
With `ad_stack_size=0` i.e. adaptive ad stack size, it raises a stmt type mismatch error (requires `AdStackAllocaStmt` but get `LocalStmtLoad`) in `determine_ad_stack_size` of `control_flow_graph`.

**Comment by erizmr** [comment]
Fixed by #4041 .

**Comment by erizmr** [comment]
Fix by #4041 .
